src/cpu/flexcpu/simple_dataflow_cpu.hh

```diff
@@ -192,33 +178,134 @@ class SimpleDataflowCPU : public BaseCPU
 
     // BEGIN Internal constants
 
-    EventFunctionWrapper attemptAllDataXlationsEvent = EventFunctionWrapper(
-        [this]{ attemptAllDataXlations(); }, name()+"attemptAllDataXlations");
-    EventFunctionWrapper attemptAllExecutionsEvent = EventFunctionWrapper(
-        [this]{ attemptAllExecutions(); }, name()+"attemptAllExecutions");
-    EventFunctionWrapper attemptAllFetchReqsEvent = EventFunctionWrapper(
-        [this]{ attemptAllFetchReqs(); }, name()+".attemptAllFetchReqsEvent",
-        false, Event::CPU_Tick_Pri);
-    EventFunctionWrapper attemptAllInstXlationsEvent = EventFunctionWrapper(
-        [this]{ attemptAllInstXlations(); }, name()+"attemptAllInstXlations");
-    EventFunctionWrapper attemptAllMemReqsEvent = EventFunctionWrapper(
-        [this]{ attemptAllMemReqs(); }, name()+".attemptAllMemReqsEvent");
-
     // END Internal constants
 
 
     // BEGIN Internal parameters
 
-    bool clockedDtbTranslation;
-    bool clockedExecution;
-    bool clockedInstFetch;
-    bool clockedItbTranslation;
-    bool clockedMemoryRequest;
-
     Cycles executionLatency;
 
     // END Internal parameters
 
+    /**
+     * Models a functional unit. Each unit has a latency, in cycles, and a
+     * bandwidth in accesses per cycle. A latency of 0 implies the action is
+     * immediately taken. Otherwise, the action of the functional unit is not
+     * functionally "executed" until after latency cycles. If the functional
+     * unit is not "clocked", then the latency is absolute. If it is clocked
+     * then the "execution" always occurs on the next clock edge after the
+     * latency.
+     *
+     * NOTE: We might want to make this a SimObject so it can be configured
+     *       from python more easily.
+     */
+    class Resource
+    {
+      protected:
+        SimpleDataflowCPU* cpu;
+
+      private:
+        const bool clocked;
+        const Cycles latency;
+        const int bandwidth;
```

Comment:
I don't think this gives us enough information to model what we want to model. I'd like to try to draw this out in person.

Reply:
Like I think we mentioned it a little when discussing how to limit latency and bandwidth, but I think we also need to think about pipelinability?

Reply:
I think this part shouldn't stop us from merging the patch, since not merging now will make this very large changeset harder to merge, but it will probably look pretty different by the time we fully implement constraints in their final form.

Reply:
I agree with this :).